src/viam/trajex/totg/trajectory.cpp

```diff
@@ -370,66 +378,104 @@ enum class integration_event : std::uint8_t {
+        // Move cursor to next segment
         cursor.seek(boundary);
         auto segment_after = *cursor;
 
-        // Sample geometry from segment starting at boundary
+        // Sample geometry on RIGHT side of boundary
         const auto q_prime_after = segment_after.tangent(boundary);
         const auto q_double_prime_after = segment_after.curvature(boundary);
 
-        // Compute acceleration limit curve (s_dot_max_acc) on both sides of boundary
-        const auto [s_dot_max_acc_before, s_dot_max_vel_before] =
+        // Compute s_dot_max_acc on both sides
+        const auto [s_dot_max_acc_before, _1] =
             compute_velocity_limits(q_prime_before, q_double_prime_before, opt.max_velocity, opt.max_acceleration, opt.epsilon);
 
-        const auto [s_dot_max_acc_after, s_dot_max_vel_after] =
+        const auto [s_dot_max_acc_after, _2] =
             compute_velocity_limits(q_prime_after, q_double_prime_after, opt.max_velocity, opt.max_acceleration, opt.epsilon);
 
-        // Switching velocity is the minimum of acceleration limit curve on both sides (equation 38)
-        const double s_dot_switching = std::min(s_dot_max_acc_before, s_dot_max_acc_after);
+        // Compute limit curve slopes using numerical approximation
+        const arc_length before_boundary = std::max(boundary - arc_length{opt.epsilon}, current_segment.start());
+        const auto q_prime_bb = current_segment.tangent(before_boundary);
+        const auto q_double_prime_bb = current_segment.curvature(before_boundary);
+        const auto [s_dot_max_acc_bb, _5] =
+            compute_velocity_limits(q_prime_bb, q_double_prime_bb, opt.max_velocity, opt.max_acceleration, opt.epsilon);
+
+        const double actual_step_left = (boundary - before_boundary).value;
+        if (actual_step_left < opt.epsilon * 0.5) {
+            continue;
+        }
+        // this is d/ds s_dot_max_acc(s-)
+        const double slope_left = (s_dot_max_acc_before - s_dot_max_acc_bb) / actual_step_left;
+
+        const arc_length after_boundary = std::min(boundary + arc_length{opt.epsilon}, segment_after.end());
+        const auto q_prime_ab = segment_after.tangent(after_boundary);
+        const auto q_double_prime_ab = segment_after.curvature(after_boundary);
+        const auto [s_dot_max_acc_ab, _6] =
+            compute_velocity_limits(q_prime_ab, q_double_prime_ab, opt.max_velocity, opt.max_acceleration, opt.epsilon);
 
-        // Validate switching point conditions:
-        // 1. Switching velocity must be within velocity limits
-        const double s_dot_max_vel_boundary = std::min(s_dot_max_vel_before, s_dot_max_vel_after);
-        if (s_dot_switching > s_dot_max_vel_boundary + opt.epsilon) {
-            continue;  // Switching velocity exceeds velocity limit
+        const double actual_step_right = (after_boundary - boundary).value;
+        if (actual_step_right < opt.epsilon * 0.5) {
+            continue;
+        }
```

Comment:
This could be done immediately after 432 and before invoking `compute_velocity_limits`.